### PR TITLE
Add test for deepseek_coder

### PR DESCRIPTION
**Pull request by meenakshiramanathan1**
Model is currently failing in verify with pcc drop of 0.32

### CODE REVIEW

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+from test.models.pytorch.multimodal.deepseek_coder.utils.load_model import (
+    DeepSeekWrapper,
+    download_model_and_tokenizer,
+    generate,
+    verify_deepseek_coder,
+)
+from test.models.utils import Framework, Source, Task, build_module_name
+
+# @pytest.mark.parametrize("variant", ["deepseek-coder-1.3b-instruct"])
```
Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py:17]: Why this is commented here?

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        framework=Framework.PYTORCH, model="deepseek", variant=variant, task=Task.QA, source=Source.HUGGINGFACE
+    )
+
+    # Record Forge Property
+    record_forge_property("model_name", module_name)
+
+    model_name = f"deepseek-ai/{variant}"
+    model, tokenizer, inputs = download_model_and_tokenizer(model_name)
+    framework_model = DeepSeekWrapper(model)
+    padded_inputs = torch.zeros(1, 276, dtype=torch.long)
```
Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py:43]: What this `276` represents here?

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    compiled_model = forge.compile(framework_model, sample_inputs=[padded_inputs], module_name=module_name)
```
Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py:44]: Shall we have the same test function structure which we followed in other model tests?

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    verify_deepseek_coder(inputs, framework_model, compiled_model)
```
Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py:45]: Can we extend the verify function's functionality for deepseek using a decorator?

File: forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    padded_inputs = torch.zeros((batch_size, max_seq_len), dtype=inputs.dtype, device=inputs.device)
+    padded_inputs[:, :seq_len] = inputs
+
+    current_pos = seq_len
+
+    for _ in range(max_new_tokens):
+        logits = model(padded_inputs)
+
+        # Get only the logits corresponding to the last valid token
+        if isinstance(logits, list):
```
Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py:32]: pls ensure we don't get any tuples from both framework_model and compiled_model.

Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py:32]: Shall we have have like this for better clarity what difference we made when logits are tuple / list?
```python
if isinstance(logits, list):
    logits = logits[0]
next_token_logits = logits[:, current_pos - 1, :]
next_token_id = torch.argmax(next_token_logits, dim=-1)
```

File: forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+def verify_deepseek_coder(inputs, framework_model, compiled_model, max_new_tokens=200):
+    batch_size, seq_len = inputs.shape
+    max_seq_len = seq_len + max_new_tokens
+
+    padded_inputs = torch.zeros((batch_size, max_seq_len), dtype=inputs.dtype, device=inputs.device)
+    padded_inputs[:, :seq_len] = inputs
+    verify([padded_inputs], framework_model, compiled_model)
+
+
+def generate(max_new_tokens, model, inputs, tokenizer):
```
Comment (ashokkumarkannan1) [inline forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py:19]: Can we have docstring for this function explaining how this function works? I'm thinking of renaming the function name to `generate_no_cache`. May I know your thoughts in this?

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py (hunk 1 above)
Comment (meenakshiramanathan1) [reply forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py]: Modified

File: forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py (hunk 6 above)
Comment (meenakshiramanathan1) [reply forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py]: Modified

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py (hunk 2 above)
Comment (meenakshiramanathan1) [reply forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py]: Changed it to a generic way now

**Review by ashokkumarkannan1 — CHANGES_REQUESTED** [review]

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py (hunk 4 above)
Comment (meenakshiramanathan1) [reply forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py]: Modified it

**Comment by ashokkumarkannan1** [comment]
> Random inputs is passed to forge compilation as when actual inputs are passed, model faces data mismatch from tvm, so currently actual inputs are passed while running compiled model which generates outputs and random inputs are passed during compilation, [issue](https://github.com/tenstorrent/tt-forge-fe/issues/1325) has been created for this. Model is currently failing in verify with pcc drop of 0.26

@nvukobratTT ,
May I know your thoughts in this on passing random inputs forge compilation alone and passing the actual input to compiled model inference. However we have the proper output generated from the compiled model.

@meenakshiramanathan1 Can you please paste some generated answer from both framework and compiled model for future reference?

File: forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    DeepSeekWrapper,
+    download_model_and_tokenizer,
+    generate_no_cache,
+    pad_inputs,
+)
+from test.models.utils import Framework, Source, Task, build_module_name
+
+
+@pytest.mark.parametrize("variant", ["deepseek-coder-1.3b-instruct"])
+def test_deepseek_inference_no_cache_cpu(variant):
```
Comment (nvukobratTT) [inline forge/test/models/pytorch/multimodal/deepseek_coder/test_deepseek_coder.py:19]: Let's put these CPU tests on end of the file. The most relevant tests should be on the top

File: forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
```
Comment (nvukobratTT) [inline forge/test/models/pytorch/multimodal/deepseek_coder/utils/load_model.py:1]: Should we rename the file name to accommodate the purpose better?

**Comment by vkovinicTT** [comment]
LGTM

**Review by nvukobratTT — APPROVED** [review]
